### PR TITLE
'Solution'

### PR DESCRIPTION
**Pull request by IvanGuculyak**

### CODE REVIEW

File: app/battle/batlle.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+from app.knight.knight import Knight
```
Comment (ArsenPidhoretskyi) [inline app/battle/batlle.py:1]: As for me, It is not good to name the file the same as the folder.
I recommend you to rename it to model or entity

File: app/battle/batlle.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+from app.knight.knight import Knight
+
+
+def battle_for_two(first_knight: Knight, second_knight: Knight) -> dict:
+    first_knight.hp = first_knight.hp - second_knight.power
+    second_knight.hp = second_knight.hp - first_knight.power
```
Comment (ArsenPidhoretskyi) [inline app/battle/batlle.py:6]: Wrong calculation:

> They lose their hp based on power of opponent and self protection using next formula:
> red_knight_stats["hp"] -= x_knight_stats["power"] - red_knight_stats["protection"]
> x_knight_stats["hp"] -= red_knight_stats["power"] - x_knight_stats["protection"]

File: app/knight/knight.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        for part in armour:
+            self.hp += part["protection"]
+        return self.hp
+
+    def weapon(self, weapon_power: int) -> int:
+        self.power += weapon_power
+        return self.power
+
+    def potion(self, potion: dict) -> callable:
+        if potion is not None:
```
Comment (ArsenPidhoretskyi) [inline app/knight/knight.py:20]: Due to your type annotations it always dict. Add None or Optional to potion type hinting

File: app/knight/knight.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    def armour(self, armour: list) -> int:
+        for part in armour:
+            self.hp += part["protection"]
+        return self.hp
+
+    def weapon(self, weapon_power: int) -> int:
+        self.power += weapon_power
+        return self.power
+
+    def potion(self, potion: dict) -> callable:
```
[the remark this answers is not in the document]
Comment (ArsenPidhoretskyi) [reply app/knight/knight.py:19]: BTW, return self.hp, self.power does not callable. It is tuple[int, int]

File: app/knight/knight_model.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+            self.protection += part["protection"]
+        return self.hp
+
+    def weapon(self, weapon_power: int) -> int:
+        self.power += weapon_power
+        return self.power
+
+    @staticmethod
+    def get_item(value_of_dict: dict, item: str) -> int:
+        return value_of_dict[item] if value_of_dict.get(item) else 0
```
Comment (ArsenPidhoretskyi) [inline app/knight/knight_model.py:29]: value_of_dict.get(item, 0) maybe?

File: app/knight/knight_model.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+            hp: int,
+            protection: int = 0
+    ) -> None:
+
+        self.name = name
+        self.power = power
+        self.hp = hp
+        self.protection = protection
+
+    def armour(self, armour: list) -> int:
```
Comment (ArsenPidhoretskyi) [inline app/knight/knight_model.py:18]: rename to apply_armour
Function name should be a verb not noun

Comment (ArsenPidhoretskyi) [reply app/knight/knight_model.py:18]: The same with weapon

File: app/battle/batlle_of_knights.py [hunk 7]
```diff
@@ -1,9 +1,9 @@
-from app.knight.knight import Knight
+from app.knight.knight_model import Knight
```
Comment (ArsenPidhoretskyi) [inline app/battle/batlle_of_knights.py:1]: Before was better 
knight > knight_model
Okay. If you need choose or `knight` or `model`)))

File: app/knight/knight_model.py (hunk 5 above)
Comment (ArsenPidhoretskyi) [reply app/knight/knight_model.py:29]: Then you can delete function)

**Review by ArsenPidhoretskyi — CHANGES_REQUESTED** [review]

**Review by Vasyl-Smutok — APPROVED** [review]
Well Done 👍

**Pull request closed without merging**